src/drivers/esp8266_drv.c

```diff
@@ -1128,38 +1206,163 @@ bool esp8266_drv_init(struct Serial *s, const int priority,
         return true;
 }
 
-static void _connect_cb(bool status, const int chan_id)
+/**
+ * Callback method invoked when a callback completes.
+ */
+static void connect_cb(const bool status, const bool already_connected)
 {
-        struct channel *ch = esp8266_state.comm.channels + chan_id;
+        struct channel_sync_op *cso = &esp8266_state.comm.connect_op;
+        struct channel *ch = esp8266_state.comm.channels + cso->chan_id;
+
         ch->connected = status;
+
+        if (!status && already_connected) {
+                /* Attempt to close the connection to rectify state */
+                pr_info(LOG_PFX "Closing unexpectedly open connection\r\n");
+                esp8266_close(cso->chan_id, NULL);
+        }
+
         cmd_set_check(CHECK_DATA);
+
+        xSemaphoreGive(cso->cb_semaphore);
 }
 
+/**
+ * Synchronous command used to open a new connection to a given destinataion
+ * @param proto The protocol to use.
+ * @param addr The destination address (either IP or DNS name).
+ * @param port The destination port to connect to.
+ * @return A valid Serial object if a connection was made, NULL otherwise.
+ */
 struct Serial* esp8266_drv_connect(const enum protocol proto,
-                                   const char* dst_ip,
-                                   const unsigned int dst_port)
+                                   const char* addr,
+                                   const unsigned int port)
 {
-        const int chan_id = get_next_free_channel_num();
-        if (chan_id < 0) {
+        /* This is synchronous code for now */
+        struct channel_sync_op *cso = &esp8266_state.comm.connect_op;
+        xSemaphoreTake(cso->op_semaphore, portMAX_DELAY);
```

Comment:
I realized the PR notes and this builds on the previous one I reviewed. holding off comments.